### DISCUSSION TITLE
replace_rcpt for outgoing email

### DISCUSSION THREAD

**Discussion by Nairou** [emersion/maddy#525] [Q&A] [answered]
Apologies for bringing this up again. I'm trying to apply the change you mentioned in #419, but am uncertain exactly how the `replace_rcpt` line should be integrated into the `local_rewrites` section. Every attempt I've made so far gives an invalid config.

**Comment by nickersonm** [comment, marked as answer]
If you put it in `local_rewrites`, then in the `authorize_sender` section the 'from' address will be transformed according to the regexp rule before checking if it's a valid sender. See [the examples here](https://maddy.email/reference/modifiers/envelope/#envelope-sender-recipient-rewriting), or alternatively use a database query [as shown here](https://maddy.email/reference/table/sql_query/).

**Reply by Nairou** [reply]
Thanks for the response. My confusion is what else I need to put in `local_rewrites` for it to accept the `replace_rcpt` line.

This is my current `local_rewrites`:

```
table.chain local_rewrites {
    optional_step regexp "(.+)\+(.+)@(.+)" "$1@$3"
    optional_step static {
        entry postmaster postmaster@$(primary_domain)
    }
    optional_step file /data/aliases
}
```

If I drop a `replace_rcpt regexp ".*" "catchall@example.com"` within that block at the end, I get an error trying to restart the server. I also tried wrapping it in a `modify{}` block, but still got an error.

**Reply by nickersonm** [reply]
Note that `local_rewrites` is a [table chain](https://maddy.email/reference/table/chain/) that's (by default) already used in a `replace_rcpt`, e.g. `replace_rcpt &local_rewrites`. You don't want `replace_rcpt` for sender address authorization, you are dealing with the `prepare_email` directive instead. So what you want to do here is either:

- Add a new `optional_step` to the end of the `local_rewrites` chain that transforms whatever's left to the catchall address, in which case the catchall address will also *receive* unaddressed emails
- Make a new `table.chain` to use with `prepare_email` that includes both `local_rewrites` and the catchall regexp steps

In either case the directive will be checking the address coming out of the table transformation against the logged in user.

I do something similar, albeit more complicated, using:

```conf
# Local address rewriting
table.chain local_rewrites {
  # Plus-address rewriting
  optional_step regexp "(.+)\+(.+)@(.+)" "$1@$3"
  
  # Defined aliases
  optional_step sql_query {
    driver sqlite3
    dsn smtp.db
    init "CREATE TABLE IF NOT EXISTS alias (source_like TEXT, source_glob TEXT, notsource TEXT, destination TEXT NOT NULL)"
    lookup "SELECT DISTINCT destination FROM alias WHERE (:key LIKE source_like OR :key GLOB source_glob) AND destination NOT IN (SELECT destination FROM alias WHERE :key GLOB notsource)"
  }
  
  step identity
}

# Allowed sending aliases
#   "domain" for any address at specified domain
#   "*" for any address at all
#   "email@domain" for specific alternate address
table.chain sendalias {
  optional_step sql_query {
    driver sqlite3
    dsn smtp.db
    init "CREATE TABLE IF NOT EXISTS sendalias (address TEXT PRIMARY KEY NOT NULL, alias TEXT NOT NULL)"
    lookup "SELECT DISTINCT alias FROM sendalias WHERE :key LIKE address"
  }
  
  step identity
}

...

# SMTP user submission server
submission tls://0.0.0.0:465 tcp://0.0.0.0:587 {
  
  ...
  
  # Process local domains
  source $(local_domains) {
    # Check that sender is authorized
    check {
      authorize_sender {
        prepare_email &local_rewrites
        user_to_email &sendalias
      }
    }
    
    ...
}
```

**Reply by Nairou** [reply]
Thank you VERY much for your help! I got the first option (add a new `optional_step` to the end of the `local_rewrites` chain) to work, and that's good enough for me, as I'm the only user on this email server. I don't know if that will bite me later, or if the config will still route to/from real addresses before the catchall.

I really wish the config made sense to me, I couldn't for the life of me get the second option to work. I don't use a SQL database, so couldn't follow your example too closely. But, I have it working well enough now, so thank you!

**Reply by nickersonm** [reply]
You're welcome! If you want to explore the configuration more, I suggest going through all the documentation a few times and also translating how you think the default configuration works. I found it helpful to rebuild the default configuration starting from scratch and adding a few lines at the time while observing behavior with `maddy run --debug`.
